gateserver/utils/structparse.py

```diff
@@ -0,0 +1,38 @@
+from collections import namedtuple
+from struct import Struct
+from enum import Enum
+
+STRUCT_FORMAT = '<' # little-endian, no alignment (i.e. packed)
+
+class t:
+    """pieces of struct format strings: docs.python.org/3/library/struct.html"""
+    uint8 = 'B'
+    bytes = lambda sz: '{}s'.format(sz)
+
+class MyStructMixin:
+    _struct = None
+
+    @classmethod
+    def set_struct(cls, formatstring):
+        cls._struct = Struct(formatstring)
+
+    @classmethod
+    def unpack_from(cls, buf):
+        """Constructs a new instance by unpacking the given buffer.
+
+        Returns the new instance and the rest of the buffer.
+        """
+        sz = cls._struct.size
+        head, tail = buf[:sz], buf[sz:]
+        return cls(*cls._struct.unpack(head)), tail
```

Comment:
Preco nie?
A ano, nie kazda struktura ma tail, ved to som aj pisal. Pre struktury bez tailu je to error. (Teraz to robis adhoc.)